crypto/x509v3/v3_alt.c

```diff
@@ -157,15 +159,36 @@ int GENERAL_NAME_print(BIO *out, GENERAL_NAME *gen)
         break;
 
     case GEN_EMAIL:
-        BIO_printf(out, "email:%s", gen->d.ia5->data);
-        break;
+	formatted_output = ASN1_STRING_new();
+	formatted_length = sizeof(gen->d.ia5->data);
+	data = calloc(7 + formatted_length, sizeof(char)); 
+	snprintf(data, 7 + formatted_length, "email:%s", gen->d.ia5->data);
+	ASN1_STRING_set(formatted_output, data, 7 + formatted_length);
+	ASN1_STRING_print(out, formatted_output);
+	ASN1_STRING_free(formatted_output);
```

Comment:
No - the approach should be a BIO_printf of the label (i.e. "email") followed by a ASN1_STRING_print.
There is no need to create a *copy* of the data and prepend a label.

GENERAL_NAME_get0_value can be used to return the ia5 value - or you can directly reference it given the current code does that. It is an ASN1_IA5STRING which is an ASN1_STRING ...

e.g. something like the following 

```
BIO_printf(out,"email:");
ASN1_STRING_print(out,gen->d.ia5);

```

Reply:
I see, I had a misunderstanding of how it worked, I thought the out BIO would have to take it from one print statement. I'll change it.